x/concentrated-liquidity/position.go

```diff
@@ -226,3 +228,101 @@ func (k Keeper) getNextPositionIdAndIncrement(ctx sdk.Context) uint64 {
 	k.SetNextPositionId(ctx, nextPositionId+1)
 	return nextPositionId
 }
+
+// fungifyChargedPosition takes in a list of positionIds and combines them into a single position.
+// The previous positions are deleted from state and the new position ID is returned.
+// An error is returned if the caller does not own all the positions, if the positions are all not fully charged, or if the positions are not all in the same pool / tick range.
+func (k Keeper) fungifyChargedPosition(ctx sdk.Context, owner sdk.AccAddress, positionIds []uint64) (uint64, error) {
+	// Check we meet the minimum number of positions to combine.
+	if len(positionIds) <= MinNumPositionsToCombine {
+		return 0, types.PositionQuantityTooLowError{MinNumPositions: MinNumPositionsToCombine, NumPositions: len(positionIds)}
+	}
+
+	// Check that all the positions are in the same pool, tick range, and are fully charged.
+	// Sum the liquidity of all the positions.
+	poolId, lowerTick, upperTick, liquidity, err := k.validatePositionsAndGetTotalLiquidity(ctx, owner, positionIds)
+	if err != nil {
+		return 0, err
+	}
+
+	// The new position's timestamp is the current block time minus the fully charged duration.
+	joinTime := ctx.BlockTime().Add(-types.FullyChargedDuration)
+
+	// Get the next position ID and increment the global counter.
+	positionId := k.getNextPositionIdAndIncrement(ctx)
+
+	// Initialize the fee accumulator for the new position.
+	if err := k.initializeFeeAccumulatorPosition(ctx, poolId, lowerTick, upperTick, positionId); err != nil {
+		return 0, err
+	}
+
+	// Check if the position already exists.
+	hasFullPosition := k.hasFullPosition(ctx, positionId)
+	if !hasFullPosition {
+		// If the position does not exist, initialize it with the provided liquidity and tick range.
+		err = k.initOrUpdatePositionUptime(ctx, poolId, liquidity, owner, lowerTick, upperTick, sdk.ZeroDec(), joinTime, positionId)
+		if err != nil {
+			return 0, err
+		}
+	} else {
+		// If the position already exists, return an error.
+		return 0, err
+	}
+
+	// Update the position in the pool based on the provided tick range and liquidity delta.
+	_, _, err = k.updatePosition(ctx, poolId, owner, lowerTick, upperTick, liquidity, joinTime, positionId)
+	if err != nil {
+		return 0, err
+	}
+
+	// Delete the previous positions.
+	for _, positionId := range positionIds {
+		err := k.deletePosition(ctx, positionId, owner, poolId)
+		if err != nil {
+			return 0, err
+		}
+	}
+
+	return positionId, nil
+}
+
+// validatePositionsAndGetTotalLiquidity checks that the positions are all in the same pool and tick range, and returns the total liquidity of the positions.
+func (k Keeper) validatePositionsAndGetTotalLiquidity(ctx sdk.Context, owner sdk.AccAddress, positionIds []uint64) (uint64, int64, int64, sdk.Dec, error) {
+	totalLiquidity := sdk.ZeroDec()
+	// Note the first position's params to use as the base for comparison.
+	basePosition, err := k.GetPosition(ctx, positionIds[0])
```

Comment:
Just to make sure, we're guaranteed that `len(positionIds) > 2` since we check it in `ValidateBasic` right?

Reply:
Yep! https://github.com/osmosis-labs/osmosis/blob/d918bccd3de6e3986c3b1f2cc24528140687cb32/x/concentrated-liquidity/types/msgs.go#L185-L187